src/main/java/com/jcraft/jsch/Session.java

```diff
@@ -2101,7 +2160,19 @@ public void disconnect() {
      * } }
      */
 
-    Channel.disconnect(this);
+    List<Channel> channelsCopy;
+    Lock l = channelsLock.writeLock();
+    l.lock();
+    try {
+      disconnectingChannels = true;
+      channelsCopy = new ArrayList<>(channels);
+    } finally {
+      l.unlock();
+    }
+
+    for (Channel c : channelsCopy) {
+      c.disconnect();
+    }
```

Comment:
I wonder if we ought to put the channel disconnect into a try / catch block to ignore any exceptions?
That way even if the call to disconnect a particular channel failed, we would continue to proceed with disconnecting the other channels and then the session itself.